### PR TITLE
Add referral system

### PR DESCRIPTION
**Pull request by lllumineux**
- "Пригласить друзей" button now works and gets referral invite link to user
- referrers now gets money award for inviting new users
- referrer_id arg was added to User db model

### CODE REVIEW

**Comment by BardRandall** [comment]
Pls use black and isort for your code

**Comment by lllumineux** [comment]
> Pls use black and isort for your code

oh my man, i'm sorry, very sorry, so sorry, sorry

**Comment by BardRandall** [comment]
I solved all conflicts, but you have to download changes and fix some problems

**Comment by lllumineux** [comment]
> I solved all conflicts, but you have to download changes and fix some problems

aight

**Comment by lllumineux** [comment]
pls approve it guys!

File: money_bot/handlers/commands_handlers.py [hunk 1]
```diff
@@ -1,13 +1,37 @@
+import re
+
 from aiogram import Dispatcher, types
 
-from money_bot.utils import markups, states
+from money_bot.example_config import INVITE_REWARD_MONEY_AMOUNT
```
Comment (BardRandall) [inline money_bot/handlers/commands_handlers.py]: Don't import example config, use try-except construction

File: money_bot/handlers/invite_handlers.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
+from aiogram import Bot, Dispatcher, types
+
+from money_bot.utils import models
+from money_bot.utils.states import GlobalStates
+
+
+async def entry_point(message: types.Message):
+    bot = await Bot.get_current().me
+    current_user = await models.User.find_one({"user_id": message.from_user.id})
+    referral_link = f"https://t.me/{bot.username}?start=referrer_id_{current_user.user_id}"
```
Comment (BardRandall) [inline money_bot/handlers/invite_handlers.py]: Add this string to strings.py

File: money_bot/utils/db_utils.py [hunk 3]
```diff
@@ -1,6 +1,22 @@
 from money_bot.utils import models
 
 
+async def is_user_in_db(user_id: int):
+    return bool(await get_user_by_id(user_id))
```
Comment (BardRandall) [inline money_bot/utils/db_utils.py]: Use get_user_by_id is None, I think it's better

File: money_bot/utils/db_utils.py [hunk 4]
```diff
@@ -1,6 +1,22 @@
...
+
+
+async def set_referrer_id(user_id: int, referrer_id: int):
+    user = await get_user_by_id(user_id)
+    user.referrer_id = referrer_id
+    await user.commit()
+
+
+async def increase_money_amount(user_id: int, increase_value_amount: int):
```
Comment (BardRandall) [inline money_bot/utils/db_utils.py:14]: So if you added this function, you should use it everywhere, even in earn module

File: money_bot/utils/models.py [hunk 5]
```diff
@@ -23,6 +23,7 @@ class User(Document):
     last_name = fields.StringField(allow_none=True)
     username = fields.StringField(allow_none=True)
     current_task_id = fields.IntegerField(default=-1)
+    referrer_id = fields.IntegerField()
```
Comment (BardRandall) [inline money_bot/utils/models.py:26]: You should set default value to -1, because now you set it in code

**Review by BardRandall — CHANGES_REQUESTED** [review]

File: money_bot/handlers/commands_handlers.py (hunk 1 above)
Comment (Birdi7) [reply money_bot/handlers/commands_handlers.py]: agree

File: money_bot/utils/db_utils.py (hunk 3 above)
Comment (Birdi7) [reply money_bot/utils/db_utils.py]: agree

File: money_bot/handlers/invite_handlers.py (hunk 2 above)
Comment (Birdi7) [reply money_bot/handlers/invite_handlers.py]: agree

File: money_bot/utils/db_utils.py [hunk 6]
```diff
@@ -1,6 +1,22 @@
 from money_bot.utils import models
 
 
+async def is_user_in_db(user_id: int):
+    return bool(await get_user_by_id(user_id))
+
+
+async def set_referrer_id(user_id: int, referrer_id: int):
```
Comment (Birdi7) [inline money_bot/utils/db_utils.py:8]: Is this function actually useful? I think this code will be used only once

File: money_bot/handlers/commands_handlers.py [hunk 7]
```diff
@@ -1,13 +1,37 @@
...
+
+async def process_setting_referrer_id(user_id: int, referrer_id: int):
+    if user_id != referrer_id:
+        current_user = await db_utils.get_user_by_id(user_id)
+        if await db_utils.is_user_in_db(referrer_id):
+            if not current_user.referrer_id:
+                await db_utils.increase_money_amount(referrer_id, INVITE_REWARD_MONEY_AMOUNT)
+                await db_utils.set_referrer_id(user_id, referrer_id)
+                return
+    await db_utils.set_referrer_id(user_id, -1)
```
Comment (Birdi7) [inline money_bot/handlers/commands_handlers.py:26]: Delete it and add the default value to the ODM schema, like @BardRandall suggested

File: money_bot/handlers/commands_handlers.py [hunk 8]
```diff
@@ -1,13 +1,37 @@
...
 
 
 async def cmd_start(message: types.Message):
-    await states.GlobalStates.main_menu_ans.set()
+    referrer_id = get_referrer_id(message.text)
+    if referrer_id:
+        await process_setting_referrer_id(message.from_user.id, referrer_id)
+    else:
+        await db_utils.set_referrer_id(message.from_user.id, -1)
```
Comment (Birdi7) [inline money_bot/handlers/commands_handlers.py:34]: The same comment about setting the default value in the User model schema

**Review by Birdi7 — CHANGES_REQUESTED** [review]
some changes are needed, overall is fine

File: money_bot/utils/models.py (hunk 5 above)
Comment (lllumineux) [reply money_bot/utils/models.py:26]: no, it's part of the structure, i've already explained u in real life

File: money_bot/utils/db_utils.py (hunk 6 above)
Comment (lllumineux) [reply money_bot/utils/db_utils.py:8]: it's just to make code clearer

File: money_bot/handlers/commands_handlers.py (hunk 7 above)
Comment (lllumineux) [reply money_bot/handlers/commands_handlers.py:26]: ya emu tolko chto obyasnil i on soglasen chto eto needed here, tam nelzya po drugomu, zavtra obyasnyu, approve pozhaluysta

Comment (lllumineux) [reply money_bot/handlers/commands_handlers.py:26]: ya hochu spat(((

File: money_bot/handlers/commands_handlers.py (hunk 8 above)
Comment (lllumineux) [reply money_bot/handlers/commands_handlers.py:34]: the same answer

**Comment by BardRandall** [comment]
@lllumineux will do default value a little bit later

**Review by BardRandall — DISMISSED** [review]

**Review by BardRandall — DISMISSED** [review]
I'll hope you fix everything

**Review by Birdi7 — DISMISSED** [review]

**Pull request merged**